server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java

```diff
@@ -0,0 +1,334 @@
+    IBP_0_10_2_IV0(-1),
+
+    // KIP-98 (idempotent and transactional producer support)
+    IBP_0_11_0_IV0(-1),
+
+    // introduced DeleteRecordsRequest v0 and FetchRequest v4 in KIP-107
+    IBP_0_11_0_IV1(-1),
+
+    // Introduced leader epoch fetches to the replica fetcher via KIP-101
+    IBP_0_11_0_IV2(-1),
+
+    // Introduced LeaderAndIsrRequest V1, UpdateMetadataRequest V4 and FetchRequest V6 via KIP-112
+    IBP_1_0_IV0(-1),
+
+    // Introduced DeleteGroupsRequest V0 via KIP-229, plus KIP-227 incremental fetch requests,
+    // and KafkaStorageException for fetch requests.
+    IBP_1_1_IV0(-1),
+
+    // Introduced OffsetsForLeaderEpochRequest V1 via KIP-279 (Fix log divergence between leader and follower after fast leader fail over)
+    IBP_2_0_IV0(-1),
+
+    // Several request versions were bumped due to KIP-219 (Improve quota communication)
+    IBP_2_0_IV1(-1),
+
+    // Introduced new schemas for group offset (v2) and group metadata (v2) (KIP-211)
+    IBP_2_1_IV0(-1),
+
+    // New Fetch, OffsetsForLeaderEpoch, and ListOffsets schemas (KIP-320)
+    IBP_2_1_IV1(-1),
+
+    // Support ZStandard Compression Codec (KIP-110)
+    IBP_2_1_IV2(-1),
+
+    // Introduced broker generation (KIP-380), and
+    // LeaderAdnIsrRequest V2, UpdateMetadataRequest V5, StopReplicaRequest V1
+    IBP_2_2_IV0(-1),
+
+    // New error code for ListOffsets when a new leader is lagging behind former HW (KIP-207)
+    IBP_2_2_IV1(-1),
+
+    // Introduced static membership.
+    IBP_2_3_IV0(-1),
+
+    // Add rack_id to FetchRequest, preferred_read_replica to FetchResponse, and replica_id to OffsetsForLeaderRequest
+    IBP_2_3_IV1(-1),
+
+    // Add adding_replicas and removing_replicas fields to LeaderAndIsrRequest
+    IBP_2_4_IV0(-1),
+
+    // Flexible version support in inter-broker APIs
+    IBP_2_4_IV1(-1),
+
+    // No new APIs, equivalent to 2.4-IV1
+    IBP_2_5_IV0(-1),
+
+    // Introduced StopReplicaRequest V3 containing the leader epoch for each partition (KIP-570)
+    IBP_2_6_IV0(-1),
+
+    // Introduced feature versioning support (KIP-584)
+    IBP_2_7_IV0(-1),
+
+    // Bup Fetch protocol for Raft protocol (KIP-595)
+    IBP_2_7_IV1(-1),
+
+    // Introduced AlterPartition (KIP-497)
+    IBP_2_7_IV2(-1),
+
+    // Flexible versioning on ListOffsets, WriteTxnMarkers and OffsetsForLeaderEpoch. Also adds topic IDs (KIP-516)
+    IBP_2_8_IV0(-1),
+
+    // Introduced topic IDs to LeaderAndIsr and UpdateMetadata requests/responses (KIP-516)
+    IBP_2_8_IV1(-1),
+
+    // Introduce AllocateProducerIds (KIP-730)
+    IBP_3_0_IV0(1),
+
+    // Introduce ListOffsets V7 which supports listing offsets by max timestamp (KIP-734)
+    // Assume message format version is 3.0 (KIP-724)
+    IBP_3_0_IV1(2),
+
+    // Adds topic IDs to Fetch requests/responses (KIP-516)
+    IBP_3_1_IV0(3),
+
+    // Support for leader recovery for unclean leader election (KIP-704)
+    IBP_3_2_IV0(4),
+
+    IBP_3_3_IV0(5);
```

Comment:
How would we define a MetadataVersion? And will there be a time where both IBP and MetadataVersion exist? How would they be compared against each other?

Reply:
While we're still supporting ZK and IBP, we decided (in the [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-778%3A+KRaft+Upgrades#KIP778:KRaftUpgrades-Compatibility,Deprecation,andMigrationPlan)) to increase the `metadata.version` any time there is an IBP bump, as well as the inverse. For example, if we make a `metadata.version` 6 during AK 3.3 development, we will add `IBP_3_3_IV1(6)`. 

After we drop ZK support, we will probably change the naming convention to no longer reference IBP. We'll probably get into more detail on that in a future KIP.

Reply:
Can we update the class documentation to be up to date with the new way of doing things?

Reply:
I filed https://issues.apache.org/jira/browse/KAFKA-13867 for JavaDoc improvements to `MetadataVersion.java`